lib/Dialect/ModArith/IR/ModArithDialect.cpp

```diff
@@ -135,52 +127,38 @@ LogicalResult BarrettReduceOp::verify() {
 }
 
 ParseResult ConstantOp::parse(OpAsmParser &parser, OperationState &result) {
-  APInt parsedValue(64, 0);
+  APInt parsedInt;
   Type parsedType;
 
-  if (failed(parser.parseInteger(parsedValue))) {
-    parser.emitError(parser.getCurrentLocation(),
-                     "found invalid integer value");
+  if (parser.parseInteger(parsedInt) || parser.parseColonType(parsedType))
     return failure();
-  }
 
-  if (parser.parseColon() || parser.parseType(parsedType)) return failure();
-
-  auto modArithType = dyn_cast<ModArithType>(parsedType);
-  if (!modArithType) return failure();
+  result.addAttribute(
+      "value", IntegerAttr::get(IntegerType::get(parser.getContext(),
+                                                 parsedInt.getBitWidth()),
+                                parsedInt));
+  result.addTypes(parsedType);
+  return success();
+}
 
-  auto outputBitWidth =
-      modArithType.getModulus().getType().getIntOrFloatBitWidth();
-  if (parsedValue.getActiveBits() > outputBitWidth)
-    return parser.emitError(parser.getCurrentLocation(),
-                            "constant value is too large for the modulus");
+LogicalResult ConstantOp::verify() {
+  auto valueBW = getValue().getBitWidth();
+  auto modBW = getType().getModulus().getValue().getBitWidth();
+  // Note: comparison will crash if the bitwidths aren't the same
+  if (valueBW > modBW ||
+      (valueBW == modBW && !getValue().ult(getType().getModulus().getValue())))
+    return emitOpError("Constant value must be smaller than modulus.");
```

Comment:
I recall discussing this at length a while back, and I don't think it makes sense to enforce this at parse time. In particular, negative attributes would bypass this check, and we have cases where `-1` is the most convenient value to provide to a `constant` op.

Instead, I think it would be easier to allow any integer as input, and let canonicalization or other ops like `mod_arith.reduce` handle converting a value to its canonical representative in the ring.